### ISSUE TITLE
Support running in a browser

### ISSUE THREAD

**Issue by matanlurey** [sass/dart-sass#25]
This would enable in-browser use, use in a service worker, etc.

**Comment by nex3** [comment]
We already isolate our `dart:io` dependency behind a configured import. I'm retargeting this to focus on adding a web-compatible implementation of the IO APIs we use.

This is something that would be nice to have, but I don't think I'm going to have time to implement or maintain it myself.

**Comment by matanlurey** [comment]
I think just having the capability would be enough, for now. A sass_browser package could always implement the gritty bits.

**Comment by nex3** [comment]
Since this is handled by a configured import, we can't really expose hooks for an external package. It has to be done in our codebase, since we handle the JS compilation.

**Comment by MGageMorgan** [comment]
Would love to make SASS available natively to shared hosting, even if slow, by using a JS version in the site's files. DreamHost shared doesn't (easily, if at all) provide Ruby and has a cap on processes running, so nodeJS won't fly. 

I love SASS's simplistic syntax and would love to run/compile it "natively" from my website, instead of uploading CSS files. Also, I don't know enough about Dart/node/JS to do this myself, otherwise I would. I'm struggling to figure out whether to learn Dart or Go, Dart seems to be more tempting PR/marketing-wise. 

Oh, and as a student I don't have the resources to pay for a VPS or anything above, as I'm lucky to have shared hosting to begin with. 

I'm really tempted to learn Dart to see if this is possible...

**Comment by yatsyk** [comment]
I think that one option could be if this module will ignore file parameter in compile api and use only data if there is no file system available as in case of browser.

There is also a drop-in replacement of node file system api for browsers: https://github.com/jvilk/BrowserFS . But even after enabling this module for fs replacement in webpack sass npm is not usable due to other globals misses in browser. I have no previous experience with dart: is it something like target (node or browser) we could change to fix result js? I've not found any information for npm build in readme.

**Comment by nex3** [comment]
@yatsyk It's more complicated than just dropping in a filesystem implementation. We'd probably want to do a separate compilation that targets the web. You'd need to start by adding a web-compatible implementation of [the IO functions](https://github.com/sass/dart-sass/tree/master/lib/src/io), but it's also not clear that the `render()` API would work great on the web—the importer API strongly assumes it's working with file paths, and it returns a typed array that may be difficult to work with in a browser context.

**Comment by jathak** [comment]
In terms of what sort of importers would be necessary for the web, I can think of two possibilities:

- an AsyncImporter that just fetches source files from URLs
- a non-async importer that just sort of fakes a filesystem with a map from filenames to source file contents

Do both of these sound reasonable? Are there others we'd want to support?

Also, is the goal here just to have a Dart library that can be imported on the web, or would we also want to distribute a compiled JS library that could be depended on by non-Dart web apps?

**Comment by Loilo** [comment]
Based on @jathak's suggestion, I could also imagine an implementation agnostic file system abstraction "interface" that allows for asynchronous answers to request. On top of that, Sass code could be stored, fetched or even generated on the fly.

**Comment by nex3** [comment]
> In terms of what sort of importers would be necessary for the web, I can think of two possibilities:
> 
> * an AsyncImporter that just fetches source files from URLs
> * a non-async importer that just sort of fakes a filesystem with a map from filenames to source file contents
> 
> Do both of these sound reasonable? Are there others we'd want to support?

I think for starters, it's fine to just do this without any importers at all, and leave it up to end-users to write importers if they want them. If we do add first-party importers, I think they should probably be distributed separately from Dart Sass itself. Running client-side is already kind of a niche feature, and I don't want to devote a ton of API surface in this package to it.

> Also, is the goal here just to have a Dart library that can be imported on the web, or would we also want to distribute a compiled JS library that could be depended on by non-Dart web apps?

Ideally both. For example, it would be super cool to be able to make examples on the Sass site editable by users, and that would be a lot easier if it was available as just a chunk of JS. But starting with Dart support is probably a good idea.

**Comment by TheComputerM** [comment]
Has in-browser use been added?

**Comment by ozanakman** [comment]
I would love to support and give some time to make this possible if you can show me how it should be done.

**Comment by nex3** [comment]
See https://github.com/sass/dart-sass/issues/25#issuecomment-376348604 for an overview of how to get started.

**Comment by initplatform** [comment]
It seems like renderSync could be run in the browser, if it relied solely on a custom importer to resolve all the file paths, used with a giant map of pathNames to scss contents. 

example: 

data.json

```json
"files": [
        {
            "filePath": "test.scss",
            "scss": "@import 'demo';"
        },
        {
            "filePath": "demo.scss",
            "scss": "SCSS_FILE_CONTENTS"
        },
```

```javascript
renderSync({
    data: "@import 'test",
    importer: (url, prev, done) => {
        const foundIndex = data.files.findIndex(file => file.filePath === url);
        return {
            file: data.files[foundIndex].filePath,
            contents: data.files[foundIndex].scss,
        }
    },
})
```

The problem I hit was I can't just import renderSync.

When importing `node_modules/sass/sass.dart.js`

```
self.chokidar = require("chokidar");
self.readline = require("readline");
self.fs = require("fs");
```

This assumes I'm running in node and not the browser.

**Comment by TheRakeshPurohit** [comment]
./node_modules/sass/sass.dart.js
Module not found: Can't resolve 'readline' in '..t/node_modules/sass'

**Comment by TheComputerM** [comment]
Any updates on this?

**Comment by bruno12mota** [comment]
Did some experimentations here, removed the require on chokidar and readline so that it imports well in a browser env. Tried using `renderSync`  like @initplatform mentioned but getting some Illegal Invocations errors which probably has something to do with using call on system functions

**Comment by bruno12mota** [comment]
A different approach would be to compile to wasm and have a small js interface to communicate with it, similar to how libs like esbuild did https://esbuild.github.io/getting-started/#wasm

**Comment by nex3** [comment]
@markwhitfeld Now that https://github.com/google/dart_cli_pkg/pull/91 has landed, the ball's in your court here. You can use that to set the `fs`, `readline`, and `chokidar` requires to `null`, but you'll also need to update the JS interop in `lib/src/io/node.dart` to gracefully degrade when `fs` isn't available and update `lib/src/node/utils.dart` to use a typed array rather than a buffer in a non-Node context.

**Comment by nex3** [comment]
@markwhitfeld Sorry for the delay! I'd love to avoid having an `optionalDependency` given that we want to discourage people from using the legacy API in the first place... is there a way we can say that it's a plain `ArrayBuffer` on the web? Or even just say that only the new compile API is supported for the web?

**Comment by nex3** [comment]
@markwhitfeld Are you still interested in working on this?

**Comment by markwhitfeld** [comment]
@nex3 Yes I am. Apologies for the radio silence, my focus switched in the beginning of November to adding support for Angular 13 in StackBlitz. It is now done, so I plan to get back on to this soon. I may reach out for some guidance on the dart side of things. I'll investigate the `ArrayBuffer` option.

**Comment by nex3** [comment]
We've just launched our new JS API, so at this point it's probably a good idea to focus on that for JS compatibility. That API doesn't use *anything* Node-specific, so it should be easier.

**Comment by TheRakeshPurohit** [comment]
> We've just launched our new JS API, so at this point it's probably a good idea to focus on that for JS compatibility. That API doesn't use _anything_ Node-specific, so it should be easier.

can you mention the API?

**Comment by markwhitfeld** [comment]
@nex3 This is great news! So, what does this mean for browser support?
To be clear, does it not use `Buffer` or any filesystem Node APIs?
If there is nothing that is node specific then would this mean that this should "just work™"?

Or is there an aspect of compatibility that may still need to be provided?
Happy to hop onto a call to flesh this out if needed.

**Comment by nex3** [comment]
The API is compatible, but the infrastructure isn't necessarily. We'll still need to factor out the internal components that load or use Node-specific libraries, provide useful error messages if the user tries to load files from a filesystem on the web, and so on.

**Comment by Pantura** [comment]
With some selected options, the only thing that prevented me from running in service worker was SASS_PATH calculation.

```
// Node.js specific exports, check to see if they exist & or polyfilled

if (typeof process !== "undefined") {
  self.process = process;
} else {
  self.process = {
    env: {
      SASS_PATH: null
    }
  }
}
```

```
const compilationResult = sass.compileString(scss, {
            logger: Logger.silent,
            alertColor: false,
            sourceMap: false,
            quietDeps: true,
            functions: null,
            sourceMapIncludeSources: false,
            importers: null

        });
```

**Comment by Pantura** [comment]
Could the getEnvironmentVariable environment be made optional?
https://github.com/sass/dart-sass/blob/55cb4fd5097975dc8fe11fa423ac9c1520c4f950/lib/src/io/vm.dart#L91

**Comment by Pantura** [comment]
Some advances. Created an issue to node_preamble https://github.com/mbullington/node_preamble.dart/issues/28 to fix an issue with wrapped global object that breaks access to self/window.location.

**Comment by curran** [comment]
I'm interested in taking a stab at this. Does anyone have a WIP repo to share that gets half way there?

I've done a browser build for ShareDB that has worked out well, maybe some techniques from there can be leveraged https://github.com/vizhub-core/sharedb-client-browser

Notable tools I'm thinking of trying:

 * [Rollup](https://rollupjs.org/guide/en/)
 * `@rollup/plugin-node-resolve`
 * `@rollup/plugin-commonjs`
 * `rollup-plugin-polyfill-node` if any Node requires remain

**Comment by jimmywarting** [comment]
Just tried the custom file importer but it's badly documented with very few examples.

I tried running this in the browser, to see if any console log would appear
```js
var {default: sass} = await import('https://jspm.dev/sass')
sass.compile('style.scss', {
  // An importer for URLs like `bgcolor:orange` that generates a
  // stylesheet with the given background color.
  importers: [{
    findFileUrl(url) {
      console.log(url)
    },
    canonicalize(url) {
      console.log(url)
    },
    load(canonicalUrl) {
      console.log(url)
    }
  }]
})
```
but all i was faced with was 
```
sass.dart.js:18277 Uncaught TypeError: Cannot read properties of undefined (reading 'isTTY')
    at StaticClosure.compile0 (sass.dart.js:18276:65)
    at Object.Primitives_applyFunction (sass.dart.js:1313:54)
    at Object.Function_apply (sass.dart.js:5901:18)
    at _callDartFunctionFast (sass.dart.js:24360:18)
    at Object.sass.compile (sass.dart.js:24334:20)
    at <anonymous>:2:6
```
that corsponds to this line: `color = color0 == null ? J.$eq$(self.process.stdout.isTTY, true) : color0,`

Guess it assumes that stuff is in a NodeJS env with a global process...

Would be nice if things could get less dependent on NodeJS modules to also work in browser and also in Deno

**Comment by curran** [comment]
Is there a way to stub out `self.process.stdout.isTTY` and other things like that as part of the build process?

Similar to what this does https://www.npmjs.com/package/rollup-plugin-polyfill-node

**Comment by jimmywarting** [comment]
Woop woop!  🎉 
Ready to try importers out when next version is released. maybe i can start using this then.

**Comment by curran** [comment]
This is exciting! How can we run it in a browser? Is there a working reference example to study? Thanks!

**Comment by jerivas** [comment]
See: https://github.com/sass/dart-sass#dart-sass-in-the-browser. TLDR: you only need to load the `sass` module and its dependencies. Then you can call `sass.compileString()` directly

**Comment by jimmywarting** [comment]
I want to use `importers` instead of `compileString` as i wanna be able to resolve dependencies dynamically

**Comment by jerivas** [comment]
You can pass your own importers as part of the settings object when calling `compileString` and `compileStringAsync`. Here's an example that treats imports as regular URLs and `fetch`es them:

```javascript
function canonicalize(url) {
  return new URL(url, window.location.toString())
}
async function load(canonicalUrl) {
  console.log(`Importing ${canonicalUrl} (async)`)
  const response = await fetch(canonicalUrl)
  if (!response.ok) throw new Error(`Failed to fetch ${canonicalUrl}: ${response.status} (${response.statusText})`);
  const contents = await response.text()
  return {
    contents,
    syntax: canonicalUrl.pathname.endsWith('.sass') ? 'indented' : 'scss'
  }
}
const importers = [{ canonicalize, load }]
const compiled = await sass.compileStringAsync('...', { importers })
```

And the sync version

```javascript
function canonicalize(url) {
  return new URL(url, window.location.toString())
}
function load(canonicalUrl) {
  console.log(`Importing ${canonicalUrl} (sync)`)
  const request = new XMLHttpRequest();
  request.onreadystatechange = () => {
    if (request.readyState === 4 && request.status !== 200)
      throw new Error(`Failed to fetch ${canonicalUrl}: ${request.status} (${request.statusText})`);
  };
  request.open("GET", canonicalUrl, false);
  request.send();
  return {
    contents: request.responseText,
    syntax: canonicalUrl.pathname.endsWith('.sass') ? 'indented' : 'scss'
  };
}
const importers = [{ canonicalize, load }]
const compiled = sass.compileString('...', { importers })
```

**Comment by jimmywarting** [comment]
Thx for this async demo. 
trying to play around with it by rapid prototyping in devtool. but it seems like sass really wasn't meant to be supported by any CDN 

tried with some content delivers that would be able to convert cjs to esm on the fly. but everyone seems to fail
```js
await import('https://esm.sh/sass') // fail
await import('https://cdn.jsdelivr.net/npm/sass/+esm') // fail
await import('https://unpkg.com/sass') // fail
var {default: sass} = await import('https://jspm.dev/sass')
sass.compileString('') // fail sass.dart.js:18277 Uncaught TypeError: Cannot read properties of undefined (reading 'isTTY')
```
Seems like it tries to load so many unnecessary stuff into the browser fs, chokidar, util, buffer, streams, readline which to me seems all pointless to have in a bundle

**Comment by jgerigmeyer** [comment]
@jimmywarting I don't think the version of Dart Sass with browser support has been released yet -- can you confirm @nex3?

**Comment by jimmywarting** [comment]
i would be nice to see a official ESM release that isn't in cjs in any form

**Comment by nex3** [comment]
@jgerigmeyer is correct, this will be released as part of Dart Sass 1.63.0 in a week or two. As the README indicates, it will be able to run in pure ESM mode without any bundling.
